src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,325 @@
+            ServiceBusOptions config, ServiceBusAccount serviceBusAccount, MessagingProvider messagingProvider, ILoggerFactory loggerFactory, bool singleDispatch)
+        {
+            _functionId = functionId;
+            _entityType = entityType;
+            _entityPath = entityPath;
+            _isSessionsEnabled = isSessionsEnabled;
+            _triggerExecutor = triggerExecutor;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _messagingProvider = messagingProvider;
+            _serviceBusAccount = serviceBusAccount;
+            _loggerFactory = loggerFactory;
+            _logger = loggerFactory.CreateLogger<ServiceBusListener>();
+            _receiver = CreateMessageReceiver();
+            _scaleMonitor = new Lazy<ServiceBusScaleMonitor>(() => new ServiceBusScaleMonitor(_functionId, _entityType, _entityPath, _serviceBusAccount.ConnectionString, _receiver, _loggerFactory));
+            _singleDispatch = singleDispatch;
+
+            if (_isSessionsEnabled)
+            {
+                _sessionMessageProcessor = _messagingProvider.CreateSessionMessageProcessor(_entityPath, _serviceBusAccount.ConnectionString);
+            }
+            else
+            {
+                _messageProcessor = _messagingProvider.CreateMessageProcessor(entityPath, _serviceBusAccount.ConnectionString);
+            }
+            _serviceBusOptions = config;
+        }
+
+        internal MessageReceiver Receiver => _receiver.Value;
+
+        internal IMessageSession MessageSession => _messageSession;
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            if (_started)
+            {
+                throw new InvalidOperationException("The listener has already been started.");
+            }
+
+
+
+
+            if (_singleDispatch)
+            {
+                if (_isSessionsEnabled)
+                {
+                    _clientEntity = _messagingProvider.CreateClientEntity(_entityPath, _serviceBusAccount.ConnectionString);
+                    if (_clientEntity is QueueClient queueClient)
+                    {
+                        queueClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                    }
+                    else
+                    {
+                        SubscriptionClient subscriptionClient = _clientEntity as SubscriptionClient;
+                        subscriptionClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                    }
+                }
+                else
+                {
+                    Receiver.RegisterMessageHandler(ProcessMessageAsync, _serviceBusOptions.MessageHandlerOptions);
+                }
+            }
+            else
+            {
+                StartMessageBatchReceiver(_cancellationTokenSource.Token);

```

Comment:
It looks like in the Sessions support PR from earlier this year, you removed the code we had that flowed tokens from this source down into ProcessMessage (v2 code [here](https://github.com/Azure/azure-webjobs-sdk/blob/a31126d8d2bff90e30c972b718a1b7ea42fefe81/src/Microsoft.Azure.WebJobs.ServiceBus/Listeners/ServiceBusListener.cs#L111)). You need to add that back, so cancellation works as expected.

Reply:
`RegisterMessageHandler` requires the function:
`Func<Message, CancellationToken, Task>` handler

so I created linked `CancellationToken` to listen both tokens